ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -99,16 +99,33 @@ muxStart verMpds bearer style rescb_m = do
                         | ptcl <- [minBound..maxBound]
                         , mode <- [ModeInitiator, ModeResponder] ]
 
+    mpsJob :: TVar m Int
+           -> PerMuxSharedState ptcl m
+           -> MiniProtocolDescriptions ptcl m
+           -> ptcl
+           -> m [m ()]
     mpsJob cnt pmss udesc mpdId = do
         let mpd = udesc mpdId
         w_i <- atomically newEmptyTMVar
         w_r <- atomically newEmptyTMVar
 
-        return [ mpdInitiator mpd (muxChannel pmss (AppProtocolId mpdId) ModeInitiator w_i cnt)
-                     >> mpsJobExit cnt
-               , mpdResponder mpd (muxChannel pmss (AppProtocolId mpdId) ModeResponder w_r cnt)
-                     >> mpsJobExit cnt
-               ]
+        let initiatorChannel :: Channel m BL.ByteString
+            initiatorChannel = muxChannel pmss
+                                (AppProtocolId mpdId)
+                                ModeInitiator
+                                w_i cnt
+
+            responderChannel :: Channel m BL.ByteString
+            responderChannel = muxChannel pmss
+                                (AppProtocolId mpdId)
+                                ModeResponder
+                                w_r cnt
+
+        return $ map (>> mpsJobExit cnt) $ catMaybes
+          [ mpdInitiator mpd <*> Just initiatorChannel
+                
+          , mpdResponder mpd <*> Just responderChannel
```

Comment:
That's true it's not a common idiom, but I find it easier to read than `fmap ($ a) f` (just one operator, instead of two).  Am I complaining about category theory ;) :D